Python/instrumentation.c

```diff
@@ -0,0 +1,1986 @@
+    (C_RETURN_EVENTS | (1 << PY_MONITORING_EVENT_CALL))
+
+
+static void
+instrument_all_executing_code_objects(PyInterpreterState *interp) {
+    _PyRuntimeState *runtime = &_PyRuntime;
+    HEAD_LOCK(runtime);
+    PyThreadState* ts = PyInterpreterState_ThreadHead(interp);
+    HEAD_UNLOCK(runtime);
+    while (ts) {
+        _PyInterpreterFrame *frame = ts->cframe->current_frame;
+        while (frame) {
+            if (frame->owner != FRAME_OWNED_BY_CSTACK) {
+                _Py_Instrument(frame->f_code, interp);
+            }
+            frame = frame->previous;
+        }
+        HEAD_LOCK(runtime);
+        ts = PyThreadState_Next(ts);
+        HEAD_UNLOCK(runtime);
+    }
+}
+
+static void
+set_events(_Py_Monitors *m, int tool_id, _PyMonitoringEventSet events)
+{
+    assert(0 <= tool_id && tool_id < PY_MONITORING_TOOL_IDS);
+    for (int e = 0; e < PY_MONITORING_UNGROUPED_EVENTS; e++) {
+        uint8_t *tools = &m->tools[e];
+        int val = (events >> e) & 1;
+        *tools &= ~(1 << tool_id);
+        *tools |= (val << tool_id);
+    }
+}
+
+void
+_PyMonitoring_SetEvents(int tool_id, _PyMonitoringEventSet events)
+{
+    assert(0 <= tool_id && tool_id < PY_MONITORING_TOOL_IDS);
+    PyInterpreterState *interp = _PyInterpreterState_Get();
+    assert(events < (1 << PY_MONITORING_UNGROUPED_EVENTS));
+    uint32_t existing_events = get_events(&interp->monitors, tool_id);
+    if (existing_events == events) {
+        return;
+    }
+    set_events(&interp->monitors, tool_id, events);
+    interp->monitoring_version++;
+    instrument_all_executing_code_objects(interp);
+}
+
+int
+_PyMonitoring_SetLocalEvents(PyCodeObject *code, int tool_id, _PyMonitoringEventSet events)
+{
+    assert(0 <= tool_id && tool_id < PY_MONITORING_TOOL_IDS);
+    PyInterpreterState *interp = _PyInterpreterState_Get();
+    assert(events < (1 << PY_MONITORING_UNGROUPED_EVENTS));
+    if (allocate_instrumentation_data(code)) {
+        return -1;
+    }
+    _Py_Monitors *local = &code->_co_monitoring->local_monitors;
+    uint32_t existing_events = get_events(local, tool_id);
+    if (existing_events == events) {
+        return 0;
+    }
+    set_events(local, tool_id, events);
+    if (is_version_up_to_date(code, interp)) {
+        /* Force instrumentation update */
+        code->_co_instrumentation_version = UINT64_MAX;
+    }
+    _Py_Instrument(code, interp);
+    return 0;
+}
+
+/*[clinic input]
+module monitoring
+[clinic start generated code]*/
+/*[clinic end generated code: output=da39a3ee5e6b4b0d input=37257f5987a360cf]*/
+/*[clinic end generated code]*/
+
+#include "clinic/instrumentation.c.h"
+
+static int
+check_valid_tool(int tool_id)
+{
+    if (tool_id < 0 || tool_id >= PY_INSTRUMENT_SYS_PROFILE) {
+        PyErr_Format(PyExc_ValueError, "invalid tool %d (must be between 0 and 5)", tool_id);
+        return -1;
+    }
+    return 0;
+}
+
+/*[clinic input]
+monitoring.use_tool
+
+    tool_id: int
+    name: object
```

Comment:
Why not this and save some boilerplate code in the impl function?

```suggestion
    name: str
```

Reply:
Clinic turns the `str` into a C `char *`, which isn't what I want.

Reply:
Ah, yeah you want a PyObject. You can use the `unicode` converter for that.

```diff
diff --git a/Python/instrumentation.c b/Python/instrumentation.c
index 7d4633da51..2231bea013 100644
--- a/Python/instrumentation.c
+++ b/Python/instrumentation.c
@@ -1605,22 +1605,18 @@ check_valid_tool(int tool_id)
 monitoring.use_tool_id
 
     tool_id: int
-    name: object
+    name: unicode
     /
 
 [clinic start generated code]*/
 
 static PyObject *
 monitoring_use_tool_id_impl(PyObject *module, int tool_id, PyObject *name)
-/*[clinic end generated code: output=30d76dc92b7cd653 input=ebc453761c621be1]*/
+/*[clinic end generated code: output=30d76dc92b7cd653 input=7fb02f6d320fe6d2]*/
 {
     if (check_valid_tool(tool_id))  {
         return NULL;
     }
-    if (!PyUnicode_Check(name)) {
-        PyErr_SetString(PyExc_ValueError, "tool name must be a str");
-        return NULL;
-    }
     PyInterpreterState *interp = _PyInterpreterState_Get();
     if (interp->monitoring_tool_names[tool_id] != NULL) {
         PyErr_Format(PyExc_ValueError, "tool %d is already in use", tool_id);
```